**Pandas**
---
O cliente é uma empresa que nos contratou como **Cientista de Dados** para fazer uma análise das **taxas de cancelamento** de seus planos ofertados para os clientes dessa empresa.

Nosso dever é dazer uma análise do *dataset* enviado pelo nosso cliente sobre os planos ofertados, verificar onde está a raiz do  problema, e apresentar uma solução. 

## Passo 1 
---

O primeiro passo a se fazer é instalar a biblioteca do **Pandas**, utilizado pa **Análise de Dados**  em **Ciência de Dados**. Isso é feito acessando o **Terminal** e instalando o Pandas através do **Gerenciador de Pacotes Pip** incluso no Python. Esse pacote deve ser instalado diretamente no **Ambiente Virtual.venv** criado especificamente para o projeto, através do comando `pip install pandas`.

## Passo 2
---
Após a instalação do **Pandas**, iremos importar a biblioteca e abrir o **dataset** no *notebook* para a melhor exibição de seus dados. Mas antes de digitar o código Python, altere o **kernel** do *notebook* para o ambiente virtual Python criado junt com o projeto.

Após isso, crie a célula de código, mas quando executar a célula, o **Jupyter** irá pedir a instalçao do pacote do ***ipykernel***. Prossiga com a instalçao para a visualização do resultado do código-fonte. 

Segue a célula do código abaixo:

In [2]:
# importação da biblioteca
import pandas as pd

# lê os dados da tabela
tabela = pd.read_csv("cancelamentos.csv")

# mostra dos dados da tabela no notebook
display(tabela)

CustomerID  idade    sexo  tempo_como_cliente  frequencia_uso  \
0              2.0   30.0  Female                39.0            14.0   
1              3.0   65.0  Female                49.0             1.0   
2              4.0   55.0  Female                14.0             4.0   
3              5.0   58.0    Male                38.0            21.0   
4              6.0   23.0    Male                32.0            20.0   
...            ...    ...     ...                 ...             ...   
881661    449995.0   42.0    Male                54.0            15.0   
881662    449996.0   25.0  Female                 8.0            13.0   
881663    449997.0   26.0    Male                35.0            27.0   
881664    449998.0   28.0    Male                55.0            14.0   
881665    449999.0   31.0    Male                48.0            20.0   

        ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
0                       5.0         18.0   Standard           Annual   
1                      10.0          8.0      Basic          Monthly   
2                       6.0         18.0      Basic        Quarterly   
3                       7.0          7.0   Standard          Monthly   
4                       5.0          8.0      Basic          Monthly   
...                     ...          ...        ...              ...   
881661                  1.0          3.0    Premium           Annual   
881662                  1.0         20.0    Premium           Annual   
881663                  1.0          5.0   Standard        Quarterly   
881664                  2.0          0.0   Standard        Quarterly   
881665                  1.0         14.0    Premium        Quarterly   

        total_gasto  meses_ultima_interacao  cancelou  
0            932.00                    17.0       1.0  
1            557.00                     6.0       1.0  
2            185.00                     3.0       1.0  
3            396.00                    29.0       1.0  
4            617.00                    20.0       1.0  
...             ...                     ...       ...  
881661       716.38                     8.0       0.0  
881662       745.38                     2.0       0.0  
881663       977.31                     9.0       0.0  
881664       602.55                     2.0       0.0  
881665       567.77                    21.0       0.0  

[881666 rows x 12 columns]

## Passo 3
---

Depois de puxar os dados da tabela para o notebook, faça o tratamento dos dados. Obeserve, por exemplo, que algumas colunas, como a coluna do ID é completamente inútil para a nossa análise. Portanto, iremos fazer outra célula de código para reirar essa coluna da nossa análise> Observe a célula de código abaixo:

In [4]:
tabela = tabela.drop("CustomerID", axis=1)
display(tabela)

idade    sexo  tempo_como_cliente  frequencia_uso  \
0        30.0  Female                39.0            14.0   
1        65.0  Female                49.0             1.0   
2        55.0  Female                14.0             4.0   
3        58.0    Male                38.0            21.0   
4        23.0    Male                32.0            20.0   
...       ...     ...                 ...             ...   
881661   42.0    Male                54.0            15.0   
881662   25.0  Female                 8.0            13.0   
881663   26.0    Male                35.0            27.0   
881664   28.0    Male                55.0            14.0   
881665   31.0    Male                48.0            20.0   

        ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
0                       5.0         18.0   Standard           Annual   
1                      10.0          8.0      Basic          Monthly   
2                       6.0         18.0      Basic        Quarterly   
3                       7.0          7.0   Standard          Monthly   
4                       5.0          8.0      Basic          Monthly   
...                     ...          ...        ...              ...   
881661                  1.0          3.0    Premium           Annual   
881662                  1.0         20.0    Premium           Annual   
881663                  1.0          5.0   Standard        Quarterly   
881664                  2.0          0.0   Standard        Quarterly   
881665                  1.0         14.0    Premium        Quarterly   

        total_gasto  meses_ultima_interacao  cancelou  
0            932.00                    17.0       1.0  
1            557.00                     6.0       1.0  
2            185.00                     3.0       1.0  
3            396.00                    29.0       1.0  
4            617.00                    20.0       1.0  
...             ...                     ...       ...  
881661       716.38                     8.0       0.0  
881662       745.38                     2.0       0.0  
881663       977.31                     9.0       0.0  
881664       602.55                     2.0       0.0  
881665       567.77                    21.0       0.0  

[881666 rows x 11 columns]

A análise anterior mostra que temos um total de **881666** linhas. Entretanto, a saída não mostra quantas dessas linhas estão com espaços vazios, que não poderão ser considerados pela futura análise.

Dessa forna, precisaremos contabilizar o número de linhas com algum tipo de dado, e quantas linhas possuem valor nulo.

In [5]:
display(tabela.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 881666 entries, 0 to 881665
Data columns (total 11 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   idade                   881664 non-null  float64
 1   sexo                    881664 non-null  object 
 2   tempo_como_cliente      881663 non-null  float64
 3   frequencia_uso          881663 non-null  float64
 4   ligacoes_callcenter     881664 non-null  float64
 5   dias_atraso             881664 non-null  float64
 6   assinatura              881661 non-null  object 
 7   duracao_contrato        881663 non-null  object 
 8   total_gasto             881664 non-null  float64
 9   meses_ultima_interacao  881664 non-null  float64
 10  cancelou                881664 non-null  float64
dtypes: float64(8), object(3)
memory usage: 74.0+ MB


None

O resultado acima mostra a quantidade de linhas de cada campo, contando inclusive com as linhas vazias. Repare que o número de linhas é diferente em alumas colunas. Na próxima célula de código, precisaremos eliminar as células com valor nulo:

In [6]:
tabela = tabela.dropna()
display(tabela.info())


<class 'pandas.core.frame.DataFrame'>
Index: 881659 entries, 0 to 881665
Data columns (total 11 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   idade                   881659 non-null  float64
 1   sexo                    881659 non-null  object 
 2   tempo_como_cliente      881659 non-null  float64
 3   frequencia_uso          881659 non-null  float64
 4   ligacoes_callcenter     881659 non-null  float64
 5   dias_atraso             881659 non-null  float64
 6   assinatura              881659 non-null  object 
 7   duracao_contrato        881659 non-null  object 
 8   total_gasto             881659 non-null  float64
 9   meses_ultima_interacao  881659 non-null  float64
 10  cancelou                881659 non-null  float64
dtypes: float64(8), object(3)
memory usage: 80.7+ MB


None

Observe agora que o número de linhas é igual para todas as colunas, que por sua vez, são iguais ao número total de linhas da tabela inteira. Agora que fizemos o **tratamento de dados**, vamos para o mais importante: a **Análise de Dados**. 

## Passo 4
---

Agora finalmente iremos iniciar a análise de dados propriamente dita. Começaremos analisando a taxa de cancelamento dos planos de assinatura da empresa. Verificaremos qual é a taxa de cancelament e de onde ela vem:

In [7]:
display(tabela["cancelou"].value_counts())

cancelou
1.0    499993
0.0    381666
Name: count, dtype: int64

Temos um número. A linha com valores **1.0** representam os cancelamentos. As linhas **0.0** representam os não cancelamentos. Mas para um leigo, dica difícil dizer se os valores são bons ou ruin. Portanto, vamos transformar esses números em porcentagem.

In [8]:
display(tabela["cancelou"].value_counts(normalize=True).map("{:.2%}".format))

cancelou
1.0    56.71%
0.0    43.29%
Name: proportion, dtype: object

O `.map()` atribui a lista uma operação que transforma os valores em **porcentagem**.

Veja que, de acordo com a nossa análise, mais da metade dos clientes cancelaram o serviço. O próximo passo é tentar entender de onde vem uma taxa tão alta de cancelamento.

---
**IMPORTANTE: Na parte da análise de dados, não tem uma informação correta para ser analisada logo de cara. Esse é um processo que vai tomar tempo, pois você, de fato, precisará analisar os dados e entender o que está acontecendo no seu *dataset*. Então pode ser que demore mais em alguns casos para encontrar o que procura antes de propor uma solução.**

## Passo 5
---
Precisaremos analisar por partes. Vamos começar pela duração dos contratos. No *dataset*, existem 3 tipos de contratos: **Anual**, **Trimestral** e **Mensal**. Vejamos qual a duração de cada um dos tipos de contratos:

In [9]:
display(tabela["duracao_contrato"].value_counts())
display(tabela["duracao_contrato"].value_counts(normalize=True).map("{:.2%}".format))

duracao_contrato
Annual       354395
Quarterly    353059
Monthly      174205
Name: count, dtype: int64

duracao_contrato
Annual       40.20%
Quarterly    40.04%
Monthly      19.76%
Name: proportion, dtype: object

Observamos algo interessante: os contratos anuais e trimestrais tem uma divisão quase igual, mas há uma discrepância muito maior na duração dos contratos mensais.

Vamos agrupar as informações da coluna **duracao_contrato** e depois fazer a média das informações. Para isso, iremos usar a função `groupby()` e repassar o nome da coluna nela:

In [10]:
display(tabela.groupby("duracao_contrato").mean(numeric_only=True))

idade  tempo_como_cliente  frequencia_uso  \
duracao_contrato                                                  
Annual            38.842165           31.446186       15.880213   
Monthly           41.552407           30.538555       15.499274   
Quarterly         38.830938           31.419916       15.886662   

                  ligacoes_callcenter  dias_atraso  total_gasto  \
duracao_contrato                                                  
Annual                       3.263401    12.465156   651.697738   
Monthly                      4.985649    15.007267   550.616435   
Quarterly                    3.265245    12.460863   651.427783   

                  meses_ultima_interacao  cancelou  
duracao_contrato                                    
Annual                         14.236107  0.460760  
Monthly                        15.478012  1.000000  
Quarterly                      14.234544  0.460255

Aqui temos uma informação alarmante: os contratos do tipo *Mensal* tiveram exatos **100 % DE CANCELAMENTO!**

#### O que pode ser feito:

Vamos remover as informações desse contrato e continuar analisando. Aqui vale lembrar que nem sempre que encontrar algo que seja ruim na sua análise de dados você retira e para por ai. A ideia é ir analisando até que chegue em um valor "aceitável" dentro da sua análise. Então é importante definir esse "valor aceitável" do seu objetivo para não ficar trabalhando sem ter um ponto de partida.